File: Test/Unit/GITActor.rb
describe 'GITActor' do
  describe '+actor' do
    before do
      @actor = GITActor.actor
    end

    should 'not be nil' do
      @actor.should.not.be.nil
    end
    should 'have name "User"' do
      @actor.name.should == 'User'
    end
    should 'have email based on $USER and $HOST' do
      host = NSProcessInfo.processInfo.hostName
      @actor.email.should == "#{ENV['USER']}@#{host}"
    end
  end
  describe '+actorWithName:' do
    before do
      @actor = GITActor.actorWithName("Randy Waterhouse")
    end

    should 'not be nil' do
      @actor.should.not.be.nil
    end
    should 'have name "Randy Waterhouse"' do
      @actor.name.should == 'Randy Waterhouse'
    end
    should 'have email based on $USER and $HOST' do
      host = NSProcessInfo.processInfo.hostName
      @actor.email.should == "#{ENV['USER']}@#{host}"
    end
  end
  describe '+actorWithName:email:' do
    before do
      @actor = GITActor.actorWithName("Randy Waterhouse", email:"randy.waterhouse@epiphyte.net")
    end

    should 'not be nil' do
      @actor.should.not.be.nil
    end
    should 'have name "Randy Waterhouse"' do
      @actor.name.should == 'Randy Waterhouse'
    end
    should 'have email "randy.waterhouse@epiphyte.net"' do
      @actor.email.should == "randy.waterhouse@epiphyte.net"
    end
  end
  describe '-description' do
    before do
      @actor = GITActor.actorWithName("Randy Waterhouse", email:"randy.waterhouse@epiphyte.net")
    end
    should 'format name and email' do
      @actor.description.should == "Randy Waterhouse <randy.waterhouse@epiphyte.net>"
    end
  end
end

File: Source/GITActor.m
//
//  GITActor.m
//  Git.framework
//
//  Created by Geoff Garside on 10/01/2010.
//  Copyright 2010 Geoff Garside. All rights reserved.
//

#import "GITActor.h"


@implementation GITActor

@synthesize name, email;

+ (NSString *)defaultName {
    return @"User";
}

+ (NSString *)defaultEmail {
    NSProcessInfo *pi = [NSProcessInfo processInfo];
    NSDictionary *env = [pi environment];
    return [NSString stringWithFormat:@"%@@%@", [env objectForKey:@"USER"], [pi hostName]];
}

+ (GITActor *)actor {
    return [[[self alloc] initWithName:[self defaultName] email:[self defaultEmail]] autorelease];
}

+ (GITActor *)actorWithName: (NSString *)name {
    return [[[self alloc] initWithName:name] autorelease];
}

+ (GITActor *)actorWithName: (NSString *)name email: (NSString *)email {
    return [[[self alloc] initWithName:name email:email] autorelease];
}

- (id)initWithName: (NSString *)theName {
    NSString *theEmail = [[self class] defaultEmail];
    return [self initWithName:theName email:theEmail];
}

- (id)initWithName: (NSString *)theName email: (NSString *)theEmail {
    if ( ![super init] )
        return nil;

    self.name  = theName;
    self.email = theEmail;

    return self;
}

- (void)dealloc {
    self.name = nil;
    self.email = nil;
    [super dealloc];
}

- (NSString *)description {
    return [NSString stringWithFormat:@"%@ <%@>", self.name, self.email];
}

@end
